### PR TITLE
[Merged by Bors] - feat: the product formula for number fields

### PR DESCRIPTION
**Pull request by fbarroero**
In this file we prove the Product Formula for number fields: for any non-zero element `x` of a number field `K`, we have `∏|x|ᵥ=1` where the product runs over the equivalence classes of absoulte values of `K` and the `|⬝|ᵥ` are suitably normalized.

## Main Results
* `NumberField.FinitePlace.prod_eq_inv_abs_norm`: for any non-zero element `x` of a number field `K`, the product `∏|x|ᵥ` of the absolute values of `x` associated to the finite places of `K` is equal to the inverse of the norm of `x`.
* `NumberField.prod_abs_eq_one`: for any non-zero element `x` of a number field `K`, we have `∏|x|ᵥ=1` where the product runs over the equivalence classes of absoulte values of `K`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:88]:
```suggestion
  rcases IsFractionRing.div_surjective (A :=𝓞 K) x with ⟨a, b, hb, rfl⟩
```

**Comment by MichaelStollBayreuth** [comment]
I'll have a look over the weekend.

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:38]: `bijective_iff_existsUnique` is a very roundabout way to go about this. You have `maximalIdeal_mk` and `mk_maximalIdeal` which are exactly the ingredients to define
```lean
noncomputable def NumberField.FinitePlace.equivHeightOneSpectrum :
    FinitePlace K ≃ HeightOneSpectrum (𝓞 K) where
  toFun := maximalIdeal
  invFun := mk
  left_inv := mk_maximalIdeal
  right_inv := maximalIdeal_mk
```

(probably this should be in `NumberTheory.NumberField.FinitePlaces`) and then lemmas
like `maximalIdeal_injective` and `maximalIdeal_inj` aren't even really necessary because they're just `(equivHeightOneSpectrum K).injective` and `(equivHeightOneSpectrum K).injective.eq_iff`. Similarly this proof becomes
```lean
  · exact (finprod_eq_of_bijective maximalIdeal (equivHeightOneSpectrum K).bijective
      (fun w ↦ (norm_embedding_eq w (x : K)).symm))
```
We have tons of API for `equiv` so you should use it.

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:40]:
```suggestion
  refine (inv_eq_of_mul_eq_one_left ?_).symm
```
It's less confusing that way

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:43]:
```suggestion
    simp [h_x_nezero]
```
For terminal simps it seems that we're now being encouraged not to squeeze them, because it makes maintaining them harder. Nobody wants to see `not_false_eq_true` anyway :-)

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:55]:
```suggestion
  have h_subs₁ : t₁ ⊆ t₀ := by simp [norm_eq_one_iff_not_mem, t₁, t₀]
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:46]: Why? Is there another approach here?

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (kbuzzard) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:93]:
```suggestion
  have ha : a ≠ 0 := by
    rintro rfl
    simp at h_x_nezero
```

**Review by kbuzzard — COMMENTED** [review]
I didn't really look at the full argument but I feel like parts of it are more painful than they could be. On the other hand getting over the line is definitely the first goal, so thanks a lot for this important start!

**Comment by fbarroero** [comment]
> I didn't really look at the full argument but I feel like parts of it are more painful than they could be. On the other hand getting over the line is definitely the first goal, so thanks a lot for this important start!

Thanks! I am not an expert so I am very happy to learn!

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (fbarroero) [reply Mathlib/NumberTheory/NumberField/ProductFormula.lean:46]: No, this comment is there to explain what is about to happen...

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:11]:
```suggestion
number field `K`, we have `∏ |x|ᵥ=1` where the product runs over the equivalence classes of absolute
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:16]:
```suggestion
`K`, the product `∏ |x|ᵥ` of the absolute values of `x` associated to the finite places of `K` is
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:19]:
```suggestion
`∏ |x|ᵥ=1`, where the product runs over the equivalence classes of absolute values of `K`.
```
Spaces for readability. (I would also prefer an empty line below the `#` and `##` lines, but this is a matter of taste.)
Maybe say something about how the absolute value is normalized?

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:31]:
```suggestion

open Function Ideal IsDedekindDomain HeightOneSpectrum in
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:56]:
```suggestion
    classical
    simp only [imp_false, Associates.count_ne_zero_iff_dvd h_span_nezero (irreducible v)] at hv
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:102]:
```suggestion

open FinitePlace in
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:32]: I suggest separating out the lemmas
```lean
namespace IsDedekindDomain.HeightOneSpectrum

open NumberField FinitePlace

lemma apply_equivHeightOneSpectrum_symm (v : HeightOneSpectrum (𝓞 K)) (x : K) :
    (equivHeightOneSpectrum.symm v) x = ‖embedding v x‖ := by
  have : v = (equivHeightOneSpectrum.symm v).maximalIdeal := by
    show v = equivHeightOneSpectrum (equivHeightOneSpectrum.symm v)
    exact (Equiv.apply_symm_apply _ v).symm
  convert (norm_embedding_eq (equivHeightOneSpectrum.symm v) x).symm

open Ideal in
lemma embedding_mul_absNorm (v : HeightOneSpectrum (𝓞 K)) {x : 𝓞 K} (h_x_nezero : x ≠ 0) :
    ‖(embedding v) ↑x‖ * absNorm (v.maxPowDividing (span {x})) = 1 := by
  rw [maxPowDividing, map_pow, Nat.cast_pow, norm_def, vadicAbv_def,
    WithZeroMulInt.toNNReal_neg_apply _ (v.valuation.ne_zero_iff.mpr
    (RingOfIntegers.coe_ne_zero_iff.mpr h_x_nezero))]
  push_cast
  rw [← Real.rpow_natCast, ← Real.rpow_intCast, ← Real.rpow_add (mod_cast Nat.zero_lt_of_lt
    (mod_cast one_lt_norm v))]
  norm_cast
  rw [zpow_eq_one_iff_right₀ (Nat.cast_nonneg' (Nat.card (𝓞 K ⧸ v.asIdeal)))
    (by exact ne_of_gt (one_lt_norm v))]
  simp [valuation_eq_intValuationDef, intValuationDef_if_neg, h_x_nezero]

end IsDedekindDomain.HeightOneSpectrum
```
(between the `variable` and `namespace NumberField` lines). They both look like reasonably useful statements by themselves. Then the proof can be simplified according to the suggestion below.

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:101]:
```suggestion
  simp only [← finprod_comp_equiv equivHeightOneSpectrum.symm, apply_equivHeightOneSpectrum_symm]
  refine (inv_eq_of_mul_eq_one_left ?_).symm
  norm_cast
  have h_span_nezero : span {x} ≠ 0 := by simp [h_x_nezero]
  rw [Int.abs_eq_natAbs, ← absNorm_span_singleton,
    ← finprod_heightOneSpectrum_factorization h_span_nezero, Int.cast_natCast]
  let t₀ := {v : HeightOneSpectrum (𝓞 K) | x ∈ v.asIdeal}
  have h_fin₀ : t₀.Finite := by simp only [← dvd_span_singleton, finite_factors h_span_nezero, t₀]
  let t₁ := (fun v : HeightOneSpectrum (𝓞 K) ↦ ‖embedding v x‖).mulSupport
  let t₂ :=
    (fun v : HeightOneSpectrum (𝓞 K) ↦ (absNorm (v.maxPowDividing (span {x})) : ℝ)).mulSupport
  have h_fin₁ : t₁.Finite := h_fin₀.subset <| by simp [norm_eq_one_iff_not_mem, t₁, t₀]
  have h_fin₂ : t₂.Finite := by
    refine h_fin₀.subset ?_
    simp only [Set.le_eq_subset, mulSupport_subset_iff, Set.mem_setOf_eq, t₂, t₀,
      maxPowDividing, ← dvd_span_singleton]
    intro v hv
    simp only [map_pow, Nat.cast_pow, ← pow_zero (absNorm v.asIdeal : ℝ)] at hv
    replace hv := fun h ↦ hv (congrArg (HPow.hPow (absNorm v.asIdeal : ℝ)) h)
    classical
    simpa only [imp_false, Associates.count_ne_zero_iff_dvd h_span_nezero (irreducible v)] using hv
  have h_prod : (absNorm (∏ᶠ (v : HeightOneSpectrum (𝓞 K)), v.maxPowDividing (span {x})) : ℝ) =
      ∏ᶠ (v : HeightOneSpectrum (𝓞 K)), (absNorm (v.maxPowDividing (span {x})) : ℝ) :=
    ((Nat.castRingHom ℝ).toMonoidHom.comp absNorm.toMonoidHom).map_finprod_of_preimage_one
      (by simp) _
  rw [h_prod, ← finprod_mul_distrib h_fin₁ h_fin₂]
  exact finprod_eq_one_of_forall_eq_one fun v ↦ v.embedding_mul_absNorm h_x_nezero
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:81]:
```suggestion
  simp only [← finprod_comp_equiv equivHeightOneSpectrum.symm, apply_equivHeightOneSpectrum_symm]
  refine (inv_eq_of_mul_eq_one_left ?_).symm
  norm_cast
  have h_span_nezero : span {x} ≠ 0 := by simp [h_x_nezero]
  rw [Int.abs_eq_natAbs, ← absNorm_span_singleton,
    ← finprod_heightOneSpectrum_factorization h_span_nezero, Int.cast_natCast]
  let t₀ := {v : HeightOneSpectrum (𝓞 K) | x ∈ v.asIdeal}
  have h_fin₀ : t₀.Finite := by simp only [← dvd_span_singleton, finite_factors h_span_nezero, t₀]
  let t₁ := (fun v : HeightOneSpectrum (𝓞 K) ↦ ‖embedding v x‖).mulSupport
  let t₂ :=
    (fun v : HeightOneSpectrum (𝓞 K) ↦ (absNorm (v.maxPowDividing (span {x})) : ℝ)).mulSupport
  have h_fin₁ : t₁.Finite := h_fin₀.subset <| by simp [norm_eq_one_iff_not_mem, t₁, t₀]
  have h_fin₂ : t₂.Finite := by
    refine h_fin₀.subset ?_
    simp only [Set.le_eq_subset, mulSupport_subset_iff, Set.mem_setOf_eq, t₂, t₀,
      maxPowDividing, ← dvd_span_singleton]
    intro v hv
    simp only [map_pow, Nat.cast_pow, ← pow_zero (absNorm v.asIdeal : ℝ)] at hv
    replace hv := fun h ↦ hv (congrArg (HPow.hPow (absNorm v.asIdeal : ℝ)) h)
    classical
    simpa only [imp_false, Associates.count_ne_zero_iff_dvd h_span_nezero (irreducible v)] using hv
  have h_prod : (absNorm (∏ᶠ (v : HeightOneSpectrum (𝓞 K)), v.maxPowDividing (span {x})) : ℝ) =
      ∏ᶠ (v : HeightOneSpectrum (𝓞 K)), (absNorm (v.maxPowDividing (span {x})) : ℝ) :=
    ((Nat.castRingHom ℝ).toMonoidHom.comp absNorm.toMonoidHom).map_finprod_of_preimage_one
      (by simp) _
  rw [h_prod, ← finprod_mul_distrib h_fin₁ h_fin₂]
  exact finprod_eq_one_of_forall_eq_one fun v ↦ v.embedding_mul_absNorm h_x_nezero
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:83]:
```suggestion
/-- For any non-zero `x` in `K`, the product of `w x`, where `w` runs over `FinitePlace K`, is
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:95]:
```suggestion
  rw [← inv_eq_iff_eq_inv, inv_inv_div_inv, ← abs_div]
```

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:98]:
```suggestion
  have hb₀ : ((Algebra.norm ℤ) b : ℝ) ≠ 0 := by simp [hb]
  refine (eq_div_of_mul_eq hb₀ ?_).symm
```
I think this is a bit clearer. (There should be a `MonoidHom.div₀`; this would simplify the argument a bit...)

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:30]:
```suggestion
open Algebra
```
This suggestion seems to have disappeared. (The idea is to llocalize the `open` statements to where they are actually needed.)

Comment (fbarroero) [reply Mathlib/NumberTheory/NumberField/ProductFormula.lean:30]: I was actually wondering about that. I am making the changes now...
Thanks for the review! I am learning a lot!

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (MichaelStollBayreuth) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:37]: Perhaps these two lemmas actually fit better into `Mathlib/NumberTheory/NumberField/FinitePlaces.lean`?

**Review by MichaelStollBayreuth — APPROVED** [review]
This looks good to me now, but I'll wait for @riccardobrasca to comment before maintainer merging.

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:62]:
```suggestion
    classical
    refine (Associates.count_ne_zero_iff_dvd h_span_nezero (irreducible v)).1 <| fun h ↦ hv ?_
    congr
```
Seems more readable to me.

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (riccardobrasca) [inline Mathlib/NumberTheory/NumberField/ProductFormula.lean:79]: Nonterminal `simp`.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/NumberTheory/NumberField/ProductFormula.lean
Comment (fbarroero) [reply Mathlib/NumberTheory/NumberField/ProductFormula.lean:79]: Why is it nonterminal? is it because of the `at h_x_nezero`?
I see that 
```lean
have ha : a ≠ 0 := by simp_all 
```
and 
```lean
have ha : a ≠ 0 := by
    simp_all only [ne_eq, div_eq_zero_iff, NoZeroSMulDivisors.algebraMap_eq_zero_iff, or_false,
      not_false_eq_true]
```
also work. Are they better? Which one should be preferred?

**Pull request closed without merging**
